pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -138,15 +151,55 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
 	return predicateMetadata
 }
 
+func (topologyPairsMaps *topologyPairsMaps) AddTopologyPair(pair topologyPair, pod *v1.Pod) {
+	found := false
+	for _, existingPod := range topologyPairsMaps.topologyPairToPods[pair] {
+		if existingPod == pod {
+			found = true
+			break
+		}
+	}
+	if !found {
+		topologyPairsMaps.topologyPairToPods[pair] = append(topologyPairsMaps.topologyPairToPods[pair], pod)
+		topologyPairsMaps.podToTopologyPairs[schedutil.GetPodFullName(pod)] = append(topologyPairsMaps.podToTopologyPairs[schedutil.GetPodFullName(pod)], pair)
+	}
+}
+
+func (topologyPairsMaps *topologyPairsMaps) RemovePod(podName string) {
+	for _, pair := range topologyPairsMaps.podToTopologyPairs[podName] {
+		for index, pod := range topologyPairsMaps.topologyPairToPods[pair] {
+			if schedutil.GetPodFullName(pod) == podName {
+				podsList := topologyPairsMaps.topologyPairToPods[pair]
+				podsList[index] = podsList[len(podsList)-1]
+				if len(podsList) <= 1 {
+					delete(topologyPairsMaps.topologyPairToPods, pair)
+				} else {
+					topologyPairsMaps.topologyPairToPods[pair] = podsList[:len(podsList)-1]
+				}
+				break
+			}
+		}
+	}
+	delete(topologyPairsMaps.podToTopologyPairs, podName)
+}
+
+func (topologyPairsMaps *topologyPairsMaps) appendMaps(toAppend *topologyPairsMaps) {
+	for pod, pairs := range toAppend.podToTopologyPairs {
```

Comment:
I think adding the pairs and pods without checking if they already exists, could cause repeated values. Changing the map 'value' to a 'set' instead of a 'slice' as I suggested above will remove the concern of repeated values.